stackslib/src/net/api/get_tenures_fork_info.rs

```diff
@@ -0,0 +1,305 @@
+                    ChainError::InvalidChainstateDB
+                },
+            )?;
+            if epoch.epoch_id < StacksEpochId::Epoch30 {
+                StacksChainState::get_stacks_block_header_info_by_consensus_hash(
+                    chainstate.db(),
+                    &sn.consensus_hash,
+                )?
+                .map(|header| header.index_block_hash())
+            } else {
+                NakamotoChainState::get_nakamoto_tenure_start_block_header(
+                    chainstate.db(),
+                    &sn.consensus_hash,
+                )?
+                .map(|header| header.index_block_hash())
+            }
+        };
+        Ok(TenureForkingInfo {
+            burn_block_hash: sn.burn_header_hash.clone(),
+            burn_block_height: sn.block_height,
+            sortition_id: sn.sortition_id.clone(),
+            parent_sortition_id: sn.parent_sortition_id.clone(),
+            consensus_hash: sn.consensus_hash.clone(),
+            was_sortition: sn.sortition,
+            first_block_mined,
+        })
+    }
+}
+
+impl RPCRequestHandler for GetTenuresForkInfo {
+    /// Reset internal state
+    fn restart(&mut self) {
+        self.start_sortition = None;
+        self.stop_sortition = None;
+    }
+
+    /// Make the response
+    fn try_handle_request(
+        &mut self,
+        preamble: HttpRequestPreamble,
+        _contents: HttpRequestContents,
+        node: &mut StacksNodeState,
+    ) -> Result<(HttpResponsePreamble, HttpResponseContents), NetError> {
+        let result = node.with_node_state(|_network, sortdb, chainstate, _mempool, _rpc_args| {
+            let start_from = self
+                .stop_sortition
+                .clone()
+                .ok_or_else(|| ChainError::NoSuchBlockError)?;
+            let recurse_end = self
+                .start_sortition
+                .clone()
+                .ok_or_else(|| ChainError::NoSuchBlockError)?;
+            let recurse_end_snapshot =
+                SortitionDB::get_block_snapshot_consensus(sortdb.conn(), &recurse_end)?
+                    .ok_or_else(|| ChainError::NoSuchBlockError)?;
+            let height_bound = recurse_end_snapshot.block_height;
+
+            let mut results = vec![];
+            let mut cursor = SortitionDB::get_block_snapshot_consensus(sortdb.conn(), &start_from)?
+                .ok_or_else(|| ChainError::NoSuchBlockError)?;
+            results.push(TenureForkingInfo::from_snapshot(
+                &cursor, sortdb, chainstate,
+            )?);
+            let handle = sortdb.index_handle(&cursor.sortition_id);
+            let mut depth = 0;
+            while depth < DEPTH_LIMIT && cursor.consensus_hash != recurse_end {
+                depth += 1;
+                if height_bound >= cursor.block_height {
+                    return Err(ChainError::NotInSameFork);
+                }
+                cursor = handle.get_last_snapshot_with_sortition(cursor.block_height - 1)?;
```

Comment:
First, this needs to be `.saturating_sub()`. Otherwise, this is a runtime panic waiting to happen.

Second, you can both avoid this altogether while implementing a faster inner loop if you instead loaded the `recurse_end` snapshot, and simply used each snapshot's `parent_sortition_id` to load the parent.  Then, you can avoid the extra MARF I/O ops.

Reply:
Ah okay, then what I proposed wouldn't be sufficient.

In the future, we could put the sortition ID of the last ancestor snapshot with a sortition into each `BlockSnapshot` so we could avoid MARF reads on this path.  But that's a separate PR.

Reply:
Switched to saturating sub in 7b1ad1e